Gradual typing

skogsbaer opened the pull request:
Work done by N. Ruschmann.

Fixes #70. Fixes #322.

nruschma7 committed:
Throw error when exported function has no type spec #70

nruschma7 committed:
Move exported functions check and add command line option

nruschma7 committed:
Support and annotate functions without spec with dynamic

nruschma7 committed:
Remove debug logs

nruschma7 committed:
handle materialization

nruschma7 committed:
tally pre and postprocessing

nruschma7 committed:
Implement correct pre/postprocessig steps

albsch committed:
Support for frame variables in erlang_types.

nruschma7 committed:
Merge upstream branch 'main' into gradual-typing

nruschma7 committed:
Merge remote-tracking branch 'upstream/as/frame-variables' into gradual-typing

nruschma7 committed:
outsource tally pre/postprocessing in gradual_utils

nruschma7 committed:
simplify var constraint generation

nruschma7 committed:
handle parsing/unparsing dynamic/frame variables

nruschma7 committed:
tidy up first attempts of use case

nruschma7 committed:
simplify tally pre/postprocessing

nruschma7 committed:
add tests for gradual typing support

skogsbaer committed:
build target for unit-tests

skogsbaer committed:
fixed wording

skogsbaer committed:
re-organized gradual tests, add some more tests

Four tests are failing, all in test_files/tycheck_simple/gradual.erl

fun_mixed_fail/1
fun_mixed2_fail/1
tuple_down_fail/1
tuple_up_down_fail/1

albsch commented on src/etylizer_main.hrl:
```diff
@@ -1,16 +1,15 @@
 -type opts_mode() :: prod_mode | test_mode.
 
+% gradual_typing_mode: 'infer' (default) or 'dynamic' (default all untyped to dynamic)
```
Where is the corresponding option in the `opts` record?

albsch commented on src/gradual_utils.erl:
```diff
@@ -0,0 +1,299 @@
+-spec compose([subst:t()], subst:base_subst()) -> [subst:t()].
+compose({tally_subst, S, Fixed}, Sigma2) ->
+        S1 = apply_subst(S, Sigma2),
+        {tally_subst, S1, sets:union(Fixed, sets:from_list(maps:values(Sigma2), [{version, 2}]))}.
```
I think that's a bug and wonder why it didn't cause problems. I think the intention is to extract the names via `[N || {var, N} <- maps:values(Sigma2)]`

albsch commented on src/gradual_utils.erl:
Since the scope of variable names is very local, it's much better to not use global state here. Will change that to local counter usage.

albsch committed:
Fixed type specs

albsch committed:
tests

albsch committed:
consolidate collect_pos_neg_var

albsch committed:
Global state -> local state

albsch committed:
Remove env from exps constraint generation, it's dead

albsch committed:
dialyzer fixes

albsch committed:
merge current main

albsch committed:
Implement the feature flag to select dynamic or gradual mode.

Made gradual mode default. Exported functions with infer should have
type specs.

albsch commented:
I updated the default behavior to be in gradual mode. Infer is too slow, and it's early enough to change defaults for the tool usage.

albsch committed:
Replace subst_ty with subst.erl mechanism + manual discrimination.

albsch committed:
Use context instead of counters

albsch committed:
Test case for compose bug

albsch commented:
For future reference.

Before, we generated just subtype constraints:

```erlang
{var, L, AnyRef} ->
    Msg = utils:sformat("var ~s", pretty:render(pretty:ref(AnyRef))),
    utils:single({cvar, mk_locs(Msg, L), AnyRef, T});
```

Now, we generated two constraints, for all type variable references:

```erl
Msg = utils:sformat("var ~s", pretty:render(pretty:ref(AnyRef))),
AlphaName = fresh_ty_varname(Ctx),
Locs = mk_locs(Msg, L),
Mater = {cvarmater, Locs, AnyRef, AlphaName},
Link = {csubty, Locs, {var, AlphaName}, T},
sets:from_list([Mater, Link]);
```

The upside is we don't need to track the environment during constraint generation. The downside is that all future features have to know how to handle `cvarmater` constraints from now on.

albsch committed:
Dynamic detection for disabling redundancy checks.

* Reduced some loglevels
* Helper function for simple recursive structures
* Fixed variable naming bug
